File: Codes/2476-closest-nodes-queries-in-a-binary-search-tree.py
'''
Author: LetMeFly
Date: 2024-02-24 11:04:55
LastEditors: LetMeFly
LastEditTime: 2024-02-24 11:09:41
'''
from typing import List, Optional
from bisect import bisect_left

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def dfs(self, root: Optional[TreeNode]) -> None:
        if not root:
            return
        self.dfs(root.left)
        self.v.append(root.val)
        self.dfs(root.right)
    
    def closestNodes(self, root: TreeNode, queries: List[int]) -> List[List[int]]:
        self.v = []
        self.dfs(root)
        ans = []
        for q in queries:
            m, M = -1, -1
            loc = bisect_left(self.v, q)
            if loc != len(self.v):
                M = self.v[loc]  # v1中这里笔误写成M=loc了
                if M == q:
                    ans.append([q, q])
                    continue
            if loc:
                m = self.v[loc - 1]
            ans.append([m, M])
        return ans

File: Codes/2476-closest-nodes-queries-in-a-binary-search-tree.cpp
/*
 * @Author: LetMeFly
 * @Date: 2024-02-24 10:28:49
 * @LastEditors: LetMeFly
 * @LastEditTime: 2024-02-24 10:43:21
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> closestNodes(TreeNode* root, vector<int>& queries) {
        vector<int> v;
        queue<TreeNode*> q;
        q.push(root);
        while (q.size()) {
            TreeNode* thisNode = q.front();
            q.pop();
            v.push_back(thisNode->val);
            if (thisNode->left) {
                q.push(thisNode->left);
            }
            if (thisNode->right) {
                q.push(thisNode->right);
            }
        }
        vector<vector<int>> ans(queries.size());
        for (int i = 0; i < queries.size(); i++) {
            vector<int>::iterator upper = lower_bound(v.begin(), v.end(), queries[i]);

        }
    }
};